File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/ContentCategorySchema.java
package com.ustadmobile.lib.db.entities;

import com.ustadmobile.lib.database.annotation.UmEntity;
import com.ustadmobile.lib.database.annotation.UmPrimaryKey;
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum;
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum;

import static com.ustadmobile.lib.db.entities.ContentCategorySchema.TABLE_ID;


/**
 * Represents a schema (list) of categories.
 */
@UmEntity(tableId = TABLE_ID)
public class ContentCategorySchema {

    public static final int TABLE_ID = 2;

    @UmPrimaryKey(autoGenerateSyncable = true)
    private long contentCategorySchemaUid;

    private String schemaName;

    private String schemaUrl;

    @UmSyncLocalChangeSeqNum
    private long contentCategorySchemaLocalChangeSeqNum;

    @UmSyncMasterChangeSeqNum
    private long contentCategorySchemaMasterChangeSeqNum;


    public long getContentCategorySchemaUid() {
        return contentCategorySchemaUid;
    }

    public void setContentCategorySchemaUid(long contentCategorySchemaUid) {
        this.contentCategorySchemaUid = contentCategorySchemaUid;
    }

    public String getSchemaName() {
        return schemaName;
    }

    public void setSchemaName(String schemaName) {
        this.schemaName = schemaName;
    }

    public String getSchemaUrl() {
        return schemaUrl;
    }

    public void setSchemaUrl(String schemaUrl) {
        this.schemaUrl = schemaUrl;
    }

    public long getContentCategorySchemaLocalChangeSeqNum() {
        return contentCategorySchemaLocalChangeSeqNum;
    }

    public void setContentCategorySchemaLocalChangeSeqNum(long contentCategorySchemaLocalChangeSeqNum) {
        this.contentCategorySchemaLocalChangeSeqNum = contentCategorySchemaLocalChangeSeqNum;
    }

    public long getContentCategorySchemaMasterChangeSeqNum() {
        return contentCategorySchemaMasterChangeSeqNum;
    }

    public void setContentCategorySchemaMasterChangeSeqNum(long contentCategorySchemaMasterChangeSeqNum) {
        this.contentCategorySchemaMasterChangeSeqNum = contentCategorySchemaMasterChangeSeqNum;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ContentCategorySchema schema = (ContentCategorySchema) o;

        if (contentCategorySchemaUid != schema.contentCategorySchemaUid) return false;
        if (schemaName != null ? !schemaName.equals(schema.schemaName) : schema.schemaName != null)
            return false;
        return schemaUrl != null ? schemaUrl.equals(schema.schemaUrl) : schema.schemaUrl == null;
    }

    @Override
    public int hashCode() {
        int result = (int) (contentCategorySchemaUid ^ (contentCategorySchemaUid >>> 32));
        result = 31 * result + (schemaName != null ? schemaName.hashCode() : 0);
        result = 31 * result + (schemaUrl != null ? schemaUrl.hashCode() : 0);
        return result;
    }
}

File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/ContentCategorySchema.kt
package com.ustadmobile.lib.db.entities

import com.ustadmobile.lib.database.annotation.UmEntity
import com.ustadmobile.lib.database.annotation.UmPrimaryKey
import com.ustadmobile.lib.database.annotation.UmSyncLastChangedBy
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum
import com.ustadmobile.lib.db.entities.ContentCategorySchema.Companion.TABLE_ID



/**
 * Represents a schema (list) of categories.
 */
@UmEntity(tableId = TABLE_ID)
class ContentCategorySchema {

    @UmPrimaryKey(autoGenerateSyncable = true)
    var contentCategorySchemaUid: Long = 0

    var schemaName: String? = null

    var schemaUrl: String? = null

    @UmSyncLocalChangeSeqNum
    var contentCategorySchemaLocalChangeSeqNum: Long = 0

    @UmSyncMasterChangeSeqNum
    var contentCategorySchemaMasterChangeSeqNum: Long = 0

    @UmSyncLastChangedBy
    var contentCategorySchemaLastChangedBy: Int = 0

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val schema = other as ContentCategorySchema?

        if (contentCategorySchemaUid != schema!!.contentCategorySchemaUid) return false
        if (if (schemaName != null) schemaName != schema.schemaName else schema.schemaName != null)
            return false
        return if (schemaUrl != null) schemaUrl == schema.schemaUrl else schema.schemaUrl == null
    }

    override fun hashCode(): Int {
        var result = (contentCategorySchemaUid xor contentCategorySchemaUid.ushr(32)).toInt()
        result = 31 * result + if (schemaName != null) schemaName!!.hashCode() else 0
        result = 31 * result + if (schemaUrl != null) schemaUrl!!.hashCode() else 0
        return result
    }

    companion object {

        const val TABLE_ID = 2
    }
}
